linter/src/rules/utils.rs

```diff
@@ -4,9 +4,12 @@ use squawk_parser::ast::{
 };
 use std::collections::HashSet;
 
-pub fn tables_created_in_transaction(tree: &[RawStmt]) -> HashSet<String> {
+pub fn tables_created_in_transaction(
+    tree: &[RawStmt],
+    assume_in_transaction: bool,
+) -> HashSet<String> {
     let mut created_table_names = HashSet::new();
-    let mut inside_transaction = false;
+    let mut inside_transaction = assume_in_transaction;
```

Comment:
curious if you had thoughts on how should we handle the logic down below where we say `inside_transaction = false`?

Reply:
It's my impression that the following logic doesn't need to change with respect to `assume_in_transaction`. But there is another caveat described below.

With the external transaction, we would start with `inside_transaction = true`. Any subsequent BEGIN statements would not change `inside_transaction`. [PostgreSQL would merely issue a warning for those when executing](https://www.postgresql.org/docs/current/sql-begin.html#id-1.9.3.47.7).

Any COMMIT would then set `inside_transaction = false`, which is also correct because Postgres would commit then and warn "there is no transaction in progress" for any subsequent COMMITs.

So I think in this case, it should correctly identify tables that were created in *a* transaction in the same tree. However, if there were multiple transactions in one `tree`, it seems like we would get a set of created tables that were possibly from a different transaction. And this seems like it would happen regardless of this new `assume_in_transaction` option. Modifying a test example from  require_concurrent_index_creation.rs:

```
BEGIN;
CREATE TABLE "core_foo" (
"id" bigint GENERATED BY DEFAULT AS IDENTITY PRIMARY KEY,
"tenant_id" bigint NULL
);
COMMIT;

-- More statements/transactions ...

BEGIN;
-- Should this be flagged by the linter for not being CONCURRENTLY created, since the table was created in a different transaction?
CREATE INDEX "core_foo_tenant_id_4d397ef9" ON "core_foo" ("tenant_id");
COMMIT;
```

However, the likelihood of this being an actual production issue seems low to me. My intuition around that is, worst-case, the table is populated between the CREATE TABLE and CREATE INDEX. That leads to a slow CREATE INDEX, holding an ACCESS EXCLUSIVE lock, etc. But because these are in the same file, the deploy step executing them hasn't finished and if migrations are run before the new code is deployed, nothing is querying that new table yet.

Any thoughts on that? Is that understandable?

Reply:
Yeah that's sounds great, thank you for the thorough examples and explanation!